src/editor/tools/scenario_worker_tool.cc

```diff
@@ -0,0 +1,118 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ */
+
+#include "editor/tools/scenario_worker_tool.h"
+
+#include "editor/editorinteractive.h"
+#include "logic/map_objects/tribes/ship.h"
+#include "logic/map_objects/tribes/worker.h"
+#include "logic/player.h"
+#include "logic/widelands_geometry.h"
+
+int32_t ScenarioPlaceWorkerTool::handle_click_impl(const Widelands::NodeAndTriangle<>& center,
+                                                   EditorInteractive& eia,
+                                                   EditorActionArgs* args,
+                                                   Widelands::Map* map) {
+	if (args->new_owner < 1 || args->new_owner > map->get_nrplayers() ||
+	    args->worker_types.empty()) {
+		return 0;
+	}
+	const size_t nr_items = args->worker_types.size();
+	Widelands::EditorGameBase& egbase = eia.egbase();
+	args->infrastructure_placed = 0;
+	if (args->random_index < 0) {
+		args->random_index = std::rand() % nr_items;
+	}
+	Widelands::Player* player = egbase.get_player(args->new_owner);
+	const Widelands::WorkerDescr* wd = args->worker_types[args->random_index];
+	Widelands::Bob* bob = nullptr;
+	if (wd) {
+		Widelands::Worker& worker = wd->create(egbase, player, nullptr, center.node);
+		if (wd->becomes() != Widelands::INVALID_INDEX) {
+			worker.set_current_experience(std::max<int32_t>(
+			   0, std::min<int32_t>(args->experience, wd->get_needed_experience() - 1)));
+		}
+		if (args->carried_ware != Widelands::INVALID_INDEX) {
+			Widelands::WareInstance* wi = new Widelands::WareInstance(
+			   args->carried_ware, egbase.tribes().get_ware_descr(args->carried_ware));
+			wi->init(egbase);
+			worker.set_carried_ware(egbase, wi);
+		}
+		bob = &worker;
+	} else {
+		bob = &egbase.create_ship(center.node, player->tribe().ship(), player);
+	}
+	args->infrastructure_placed = bob->serial();
+	return 1;
+}
+
+EditorActionArgs ScenarioPlaceWorkerTool::format_args_impl(EditorInteractive& parent) {
+	EditorActionArgs a(parent);
+	a.new_owner = player_;
+	a.random_index = -1;
+	a.experience = experience_;
+	a.carried_ware = carried_ware_;
+	for (const Widelands::WorkerDescr* d : descr_) {
+		a.worker_types.push_back(d);
+	}
+	return a;
+}
+
+int32_t ScenarioDeleteWorkerTool::handle_click_impl(const Widelands::NodeAndTriangle<>& center,
+                                                    EditorInteractive& eia,
+                                                    EditorActionArgs* args,
+                                                    Widelands::Map* map) {
+	Widelands::EditorGameBase& egbase = eia.egbase();
+	Widelands::MapRegion<Widelands::Area<Widelands::FCoords>> mr(
+	   *map, Widelands::Area<Widelands::FCoords>(map->get_fcoords(center.node), args->sel_radius));
+	do {
+		std::list<Widelands::Worker*> workers_to_delete;
```

Comment:
Execute the delete code directly in the loop, then you can get rid of these 2 data structures and their secondary loops below.

Reply:
This would involve some ugly and potentially buggy iterator magic, because the linking pointers of bobs will change while deleting them. Especially as we are interested only in some bobs but not e.g. in critters. It is safer to first make a list of all items we want to delete and only do the actual deleting when all are gathered.

Reply:
Good point.